javaslang/src/main/java/javaslang/collection/Arrays.java

```diff
@@ -0,0 +1,80 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+/**
+ * Internal class, containing Java array manipulation helpers.
+ * Many arrays are represented as simple Objects, to avoid casts at the client site, and to work with all types of arrays, including primitive ones.
+ *
+ * @author Pap Lőrinc
+ * @since 2.1.0
+ */
+final class Arrays { // TODO reuse these in `Array` also
+    private static final Object[] EMPTY = {};
+    @SuppressWarnings("unchecked")
+    static <T> T[] emptyArray() { return (T[]) EMPTY; }
+
+    /** Create a single element array */
+    static Object[] asArray(Object element) {
+        return new Object[] { element };
+    }
+
+    /** System.arrayCopy with same source and destination */
+    static Object[] copyRange(Object array, int from, int to) {
+        final int length = to - from;
+        return arrayCopy(length, array, from, 0, length);
+    }
+
+    /** clone the source and set the value at the given position */
+    static Object[] copyUpdate(Object arrayObject, int index, Object element) {
+        final Object[] array = (Object[]) arrayObject;
```

Comment:
as stated in the above JavaDoc, untyped `Object` array, to get rid of call-site casts (and make room for non-Object based arrays also)